src/ResourceManagement/Resource/Microsoft.Azure.Management.V2.Resource/Authentication/AzureCredentials.cs

```diff
@@ -0,0 +1,149 @@
+                environment = environment
+            };
+            return credentials;
+        }
+
+        public static AzureCredentials FromFile(string authFile)
+        {
+            var config = new Dictionary<string, string>()
+            {
+                { "authurl", AzureEnvironment.AzureGlobalCloud.AuthenticationEndpoint },
+                { "baseurl", AzureEnvironment.AzureGlobalCloud.ResourceManagerEndpoint },
+                { "managementuri", AzureEnvironment.AzureGlobalCloud.ManagementEnpoint }
+            };
+
+            File.ReadLines(authFile)
+                .All(line =>
+                {
+                    var keyVal = line.Trim().Split(new char[] { '=' }, 2);
+                    config[keyVal[0].ToLowerInvariant()] = keyVal[1];
+                    return true;
+                });
+
+            var env = new AzureEnvironment()
+            {
+                AuthenticationEndpoint = config["authurl"].Replace("\\", ""),
+                ManagementEnpoint = config["managementuri"].Replace("\\", ""),
+                ResourceManagerEndpoint = config["baseurl"].Replace("\\", ""),
+                GraphEndpoint = config.ContainsKey("graphurl") ? config["graphurl"].Replace("\\", "") : "https://graph.windows.net"
+            };
+
+            AzureCredentials credentials = FromServicePrincipal(config["client"], config["key"], config["tenant"], env);
+            credentials.WithDefaultSubscription(config["subscription"]);
+            return credentials;
+        }
+
+        public string TenantId { get { return tenantId; } }
+
+        public string ClientId { get { return clientId; } }
+
+        public AzureCredentials withTokenCache(TokenCache tokenCache)
+        {
+            TokenCache = tokenCache;
+            return this;
+        }
+
+        public AzureCredentials WithDefaultSubscription(string subscriptionId)
+        {
+            DefaultSubscriptionId = subscriptionId;
+            return this;
+        }
+
+        public override Task ProcessHttpRequestAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var adSettings = new ActiveDirectoryServiceSettings
+            {
+                AuthenticationEndpoint = new Uri(environment.AuthenticationEndpoint),
+                TokenAudience = new Uri(environment.ResourceManagerEndpoint),
+                ValidateAuthority = true
+            };
+            string url = request.RequestUri.ToString();
+            if (url.StartsWith(environment.GraphEndpoint, StringComparison.OrdinalIgnoreCase))
+            {
+                adSettings.TokenAudience = new Uri(environment.GraphEndpoint);
+            }
+            Task<ServiceClientCredentials> credentials;
+            if (username != null && password != null)
+            {
+                credentials = UserTokenProvider.LoginSilentAsync(clientId, tenantId, username, password, adSettings, TokenCache);
+            }
```

Comment:
this way you will hit AD Endpoint for each Send and re-Authenticate. Please take a look to TestFramework authentication flow in the sdk-for-net repo.